rfcs/79-minimal-fragmenting-relationships.md

```diff
@@ -0,0 +1,126 @@
+
+Another analogy when considering the ECS as an in memory database is that relationships act as foreign keys. This unlocks a potential future where we can express more complex queries matching across multiple entities and relationship pairs. There are a myriad of additional features that can be built on top of the initial implementation described below. The main focus in this RFC will be the blocking issues that must be addressed before the initial implementation is achievable.
+## Motivation
+
+Fragmenting relationships has been a long desired bevy feature. The ability to express and traverse hierarchies and graphs within the ECS unlocks a huge amount of potential. Features such as breadth-first query traversal, immediate hierarchy clean-up, component inheritance, multi-target queries etc. all rely on or are simplified by the existence of fragmenting relationships. 
+## User-facing explanation
+
+Relationship pairs can be inserted and removed like any other component. There are 3 ids associated with any relationship pair instance:
+1. The source entity that the pair is being added to
+2. The relationship component
+3. The target entity
+
+In the minimal implementation all relationships will take the form of `(ZST Component, entity)` pairs so the API surface will look quite familiar. All naming is subject to bikeshedding:
+```rust
+#[derive(Component)]
+struct Relationship;
+
+world.entity_mut(source_entity_a)
+	.insert_pair::<Relationship>(target_entity_b)
+	.remove_pair::<Relationship>(target_entity_b);
+
+command.entity(source_entity_a)
+	.insert_pair::<Relationship>(target_entity_b)
+	.remove_pair::<Relationship>(target_entity_b);
+```
+
+Since each pair is a unique component you can also add multiple target entities with the same relationship type, as well as use APIs to access and query for them:
+```rust
+#[derive(Component)]
+struct Eats;
+
+let alice_mut = world.entity_mut(alice);
+
+alice_mut.insert_pair::<Eats>(apple)
+	.insert_pair::<Eats>(banana);
+
+alice_mut.has_pair<Eats>(apple); // == true
+
+for target in alice_mut.targets<Eats>() {
+	// target == apple
+	// target == banana
+}
+
+let query = world.query<(Entity, Targets<Eats>)>();
+for entity, targets in query.iter(&world) {
+	// entity == alice
+	for target in targets {
+		// target == apple
+		// target == banana
+	}
+}
+
+let query = world.query<Entity, With<Targets<Eats>>>();
+for entity in query.iter(&world) {
+	// entity == alice
+}
+
+```
+
+Dynamic queries would also be able to match on specific pairs:
+```rust
+let query = QueryBuilder::new::<Entity>(&mut World)
+	.with_pair::<Eats>(apple)
+	.build();
+
+query.single(&world); // == alice
+```
+
+## Implementation strategy
```

Comment:
A mention of how this interacts with intra-system parallelism is probably worth mentioning, since all of the user-facing examples are all single-threaded. A simple "this strictly relies on archetype matching and thus has no impact when used in conjunction with APIs like `Query::par_iter`" should be sufficient.